[Andrew Sutjipto] iP

over-fitted opened the pull request:
# Duke
> The minimalist ~todolist~ Duke!

## Why use todolist?
The todolist is
- **FREE** :dollar: 
- *Fast* :dash: 
- Clutter-Free! 

## Features
- [X] Managing tasks
- [X] Creating time-sensitive events and deadlines
- [ ] Setting reminders

## Installation guide
To install, follow these steps:
1. Download the latest release .jar file
2. Double click it
3. Enjoy! :tada: 

## For developers
If you [Java](https://www.java.com/en/) programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        new Duke("data/duke.txt").run();
}
```

Commit:
Add Gradle support

over-fitted committed:
Complete level 1, features echoing passed commands and exit functionality

over-fitted committed:
Add list storage and printing functionality

over-fitted committed:
Add Task class, add task completion marker and funtionality

over-fitted committed:
Add Todo, Deadilne and Event subclasses of Task

over-fitted committed:
Separate task classes out from Duke, add todo + addcustom functionality

over-fitted committed:
Add deadline adding

over-fitted committed:
Add Event functionality, fix related event-related bugs

over-fitted committed:
Fix inital regression test

over-fitted committed:
Add DukeException

over-fitted committed:
Remove non-typed tasks

over-fitted committed:
Add empty todo and misinput handling

over-fitted committed:
Convert underlying datastructre of taskList to ArrayList

over-fitted committed:
add delete functionality

over-fitted committed:
Do basic cleanup of unused items

over-fitted committed:
Delete unnecessary counter in list function

over-fitted committed:
Separate TaskList, add TaskList saving functionality

over-fitted committed:
Convert date storage type in Deadline to LocalDate type

over-fitted committed:
Merge branch 'branch-Level-7'

over-fitted committed:
Merge branch 'branch-Level-8'

over-fitted committed:
added parser and ui

over-fitted committed:
Remove data from repo

over-fitted committed:
Configure storage file name to be modifiable

over-fitted committed:
Implement OOP Storage
 and Parser

over-fitted committed:
Combine files into Duke Package

over-fitted committed:
Add Storage and Duke tests

over-fitted committed:
Expose I/O dependencies of Duke

over-fitted committed:
Rename tasklist used in parser

over-fitted committed:
Merge branch 'add-gradle-support'

over-fitted committed:
Add gradle with working tests

over-fitted committed:
Merge branch 'A-Gradle'

over-fitted committed:
Add JavaDocs

over-fitted committed:
Merge branch 'A-JavaDoc'

over-fitted committed:
Set up checkstyle with fixed style

over-fitted committed:
Merge branch 'A-CodingStandard'

over-fitted committed:
Add find functionality

over-fitted committed:
Merge branch 'branch-Level-9'

JunWei3112 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+public class Deadline extends Task {
+    private LocalDate deadline;
+
+    /**
+     * Task with a specified deadline.
+     * @param description description of task.
+     * @param deadline deadline of task.
+     */
```
Should an empty line be added between the description and the parameter section to improve readability? I noticed the same issue in other header comments too.

over-fitted replied:
Amended as of 59e8bd029f71efa19ed29b8e79276222d61587bb

JunWei3112 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Deadline extends Task {
```
Perhaps it would be better if header comments were added for this class, so that readers will be able to understand the purpose of this class at a glance? This applies to other public classes as well.

JunWei3112 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+        TaskList taskList = storage.initialise();
+
+        ui.init();
+        boolean ended = false;
```
Should a prefix be added to this variable name to make it sound more like a boolean? One example is: 
```suggestion
        boolean hasEnded = false;
```

over-fitted replied:
Amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

over-fitted replied:
amended as of 734dcb9b1359b25907ac79f9f742e0c796f93a03

JunWei3112 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String toString() {
+        String stem = super.toString();
```
Should a more descriptive variable name be used here? I noticed this in the `toString()` method in some other classes (namely Event and Deadline) too.

over-fitted replied:
As the variable names are short-lived and apparent I think the use of the name stem is appropriate

JunWei3112 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+        Pattern eventPattern = Pattern.compile("event (.*) /at (.*)");
+        Pattern findPattern = Pattern.compile("find (.*)");
+
+        // print out list
```
Perhaps it would be better if the first word of the comment was capitalised?

over-fitted replied:
Amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

JunWei3112 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,89 @@
+            return false;
+        }
+
+        if (Pattern.matches("delete \\d", input)) {
+            String[] items = input.split(" ");
+            taskList.delete(Integer.parseInt(items[1]));
+            return false;
+        }
+
+        Matcher todoMatcher = todoPattern.matcher(input);
+        if (todoMatcher.find()) {
+            taskList.addCustom(new Todo(todoMatcher.group(1)));
+            return false;
+        }
+
+        Matcher deadlineMatcher = deadlinePattern.matcher(input);
+        if (deadlineMatcher.find()) {
+            taskList.addCustom(new Deadline(deadlineMatcher.group(1), deadlineMatcher.group(2)));
+            return false;
+        }
+
+        Matcher eventMatcher = eventPattern.matcher(input);
+        if (eventMatcher.find()) {
+            taskList.addCustom(new Event(eventMatcher.group(1), eventMatcher.group(2)));
+            return false;
+        }
```
Perhaps comments can be added to each of these code blocks for consistency with the other code blocks? The reader will also be able to tell what is the function of each of these code blocks (although the code is already quite intuitive here).

over-fitted replied:
Amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

over-fitted replied:
amended as of 734dcb9b1359b25907ac79f9f742e0c796f93a03

JunWei3112 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,69 @@
+package duke;
+
+
```
Should this extra empty line here be removed?

over-fitted replied:
amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

over-fitted replied:
amended as of 734dcb9b1359b25907ac79f9f742e0c796f93a03

JunWei3112 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,69 @@
+        TaskList taskList = new TaskList();
+        try {
+            if (!this.taskListFile.createNewFile()) {
+                FileInputStream fis = new FileInputStream(this.taskListFileName);
+                ObjectInputStream ois = new ObjectInputStream(fis);
```
Perhaps it would be better if more descriptive/intuitive variable names were used here? I noticed this in the `store` method as well.

over-fitted replied:
Since the names are only used within this logical block, following the principle of short lifespan = short name I believe that these are appropriate names

JunWei3112 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+
+public class TaskList implements Serializable {
+    private ArrayList<Task> tasks;
+    private int numTask;
```
Should this variable name be changed to a more descriptive one, since readers may confuse this with the task index instead? One suggested change is
```suggestion
    private int numberOfTasks;
```

over-fitted replied:
amended as of 734dcb9b1359b25907ac79f9f742e0c796f93a03

JunWei3112 commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,95 @@
+        final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(outContent));
+
+
```
Should the extra empty line here be removed as well?

over-fitted replied:
amended as of 4030096e42b9fe4bd5c9ccecfed49e72b055c620

over-fitted replied:
amended as of 43b70d6892cfdf89357837c86596bf717b82c051

JunWei3112 reviewed (commented):
Generally, it will be better/recommended to write header comments for all public classes and methods, to allow readers to gain a better understanding of these without having to read the code? 
Also, for methods with many lines of code, it will be good to separate the logical units, and introduce each of these blocks with a comment, to improve readability.
Else, other than some minor coding standard violations to fix, LGTM! 👍🏽

over-fitted committed:
Change boolean name to fit style

over-fitted committed:
Fix documentation in Parser

over-fitted committed:
Remove unnecessary newlines

over-fitted committed:
Improve readability of task counter name in TaskList

over-fitted committed:
Amend newline formatting in javadocs

over-fitted commented:
Thank you @JunWei3112 for your constructive feedback. The style of this project has been markedly improved by your feedback

nguyiyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,95 @@
+     * @param taskList storage used to carry out actions
+     * @return whether Duke has completed executing
+     */
+    public boolean parse(String input, TaskList taskList) {
```
Perhaps the method description could be clearer and include that it checks whether the command given is a command for Duke to complete? I feel that this can help make it clearer since the method name `parse` does not give a full idea of what the method does.

over-fitted replied:
Fixed thanks!

nguyiyang commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,42 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public String toString() {
```
I think you missed out an `@Override` annotation here, this would help detect errors in case the overriding is incorrect.

over-fitted replied:
amended, thank you!

nguyiyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,95 @@
+        Pattern eventPattern = Pattern.compile("event (.*) /at (.*)");
+        Pattern findPattern = Pattern.compile("find (.*)");
+
+        // Print out list
+        if (input.equals("list")) {
+            taskList.list();
+            return false;
+        }
+
```
I like that the chunks of code are separated by a blank line followed by a comment, which makes it very clean.

nguyiyang commented on src/test/java/StorageTest.java:
```diff
@@ -0,0 +1,54 @@
+import duke.Storage;
+import duke.TaskList;
+
+public class StorageTest {
```
Perhaps a short description can be provided to explain what the tests are about.

nguyiyang reviewed (commented):
Generally, the minor coding violations mentioned have been fixed and the separation of large chunks of code makes it much cleaner. However, the parse method's functionality may not be immediately clear so I feel an additional description there would help, and descriptions for test methods would be good too. Other than that, LGTM :smiley:

over-fitted committed:
Add JavaFX tutorial to part 3

over-fitted committed:
Encapsulate I/O handling in UI

over-fitted committed:
Convert CLI to GUI

over-fitted committed:
Reorganise file, fix storage

over-fitted committed:
Merge branch 'Level-10'

over-fitted committed:
Fix style

over-fitted committed:
TaskList: Replace numOfTasks with tasks.size()

numOfTasks tracks number of items in the TaskList, which can be
trivially derived from tasks. This results in needing to maintain
an unnecessary counter which is a source for bugs.

Let's replace all instances where number of tasks is needed from using
numOfTasks to directly querying the number of tasks based on the size of
TaskList.

This results in no change to functionality while removing the space for
bugs outlined earlier.

over-fitted committed:
Handle malformed user input in Parser.

When user inputs a task ID to delete, if the input ID is not associated
with a task then Duke would throw an error where the user's input would
be ignored while the stacktrace would be printed to console.

Let's handle for such cases of malformed ID input by checking that the
ID is within the range of ID values in our list before trying to delete
the task requested.

This way we would prevent errors from being thrown by attempting to
access an item that does not exist in our tasks arraylist.

over-fitted committed:
Add equals method to TaskList

TaskList did not have an equals method, preventing other classes from
checking if their TaskList is equal to another TaskList.

Let's add an equals method to TaskList that compares that the items
within TaskList are identical to that of another TaskList.

This way Storage would be able to compare if serialised TaskList is
identical to the original TaskList input

over-fitted committed:
Add assertions

over-fitted committed:
Merge pull request #1 from over-fitted/branch-A-Assertions

Merge branch-A-Assertions

over-fitted committed:
Add override to Task.toString

over-fitted committed:
Improve documentation in Parser

over-fitted commented:
Thank you @nguyiyang for the feedback! I have fixed most of the problem areas you mentioned

over-fitted committed:
Create gradle-publish.yml

over-fitted committed:
Merge pull request #3 from over-fitted/publish-CI

Create gradle-publish.yml

over-fitted committed:
Create CS2103 default CI

over-fitted committed:
Remove buggy test

over-fitted committed:
Add update functionality

over-fitted committed:
Replace github actions checkout with v2

over-fitted committed:
Update checkout v2 step

over-fitted committed:
Merge pull request #4 from over-fitted/C-Update

C update

over-fitted committed:
Simplify todo misinput check

over-fitted committed:
Improve class documentation

over-fitted committed:
Improve class documentation

over-fitted committed:
Fix update using new items containing whitespace

over-fitted committed:
Add user guide documentation

over-fitted committed:
Implement bye command

over-fitted committed:
Update bye command documentation.

over-fitted committed:
Add Ui demo screenshot

over-fitted committed:
Set theme jekyll-theme-tactile

over-fitted committed:
Fix style and documentation

over-fitted committed:
Merge branch 'master' of github.com:over-fitted/ip

over-fitted committed:
Remove unnecessary doc files

over-fitted committed:
Add image to user guide

over-fitted committed:
Shorten deadline docs

over-fitted committed:
Fix userdocs

over-fitted committed:
Remove debugging prints

over-fitted committed:
Merge branch 'branch-A-CodeQuality'
